### DISCUSSION TITLE
Choice of μController (MIDI Host)

### DISCUSSION THREAD

**Discussion by csBlueChip** [hathach/tinyusb#1152] [Q&A]
Having just discovered [here](https://github.com/hathach/tinyusb/pull/1122#pullrequestreview-778062705) that the Pico/RP2040 has bad _Bulk Endpoint_ support and therefore _MIDI Host_ support is (essentially) impossible, I plan to take @hathach's advice and order a new batch of μC's from China.

For many reasons, I would like to move to the **ESP32-S2** platform ...Are there any known problems/limitations with the ESP32-S2 ? ...If so, what μC would you recommend?

All advice gratefully appreciated :)

**Comment by hathach** [comment]
MCU choice is up to your need, check the ci build to see which one is supported. ESP32S2 host driver is unfortunately not written just yet.

**Reply by ghost** [reply]
Hi Hathach,

Given that the supported device list is a wip, do you know, please, what mcu's have host-mode support for bulk transfer endpoints? 

Thanks

**Reply by csBlueChip** [reply]
@redteamz

Given [this article on "CI Builds"](https://docs.github.com/en/actions/automating-builds-and-tests/about-continuous-integration), I think @hathach means the [Actions tab](https://github.com/hathach/tinyusb/actions) (not the [Supported Devices page](https://github.com/hathach/tinyusb/blob/master/docs/reference/supported.rst)).

Sadly, if I'm right, there are 5,548 entries, and "*log searching is not available yet*".

**Reply by hathach** [reply]
I just check it again, those MCU with `.only.MCU_` are the one is supported https://github.com/hathach/tinyusb/tree/master/examples/host/cdc_msc_hid . So far only following mcu is supported. However, it has been a while since I tested with lpc17xx (OHCI). So basically, only rp2040 and NXP EHCI (transmidimension/chipidead) are stable. 

```
.only.MCU_LPC175X_6X
.only.MCU_LPC177X_8X
.only.MCU_LPC18XX
.only.MCU_LPC40XX
.only.MCU_LPC43XX
.only.MCU_MIMXRT10XX
.only.MCU_RP2040
```

**Reply by csBlueChip** [reply]
@redteamz You may wish to consider [this statement](https://github.com/hathach/tinyusb/pull/1122#pullrequestreview-778062705) by @hathach which states the RP2040 does NOT work with Bulk Endpoints and, due to hardware limitations, it is a BAD choice for MIDI.

The ESP32-S2 claims to contain USB hardware, BUT even [the official Espressif IDF does NOT have USB Host suport](https://esp32.com/viewtopic.php?t=16147).

So (afaict) there are no $3 μC's which work with MIDI ...and if you're going to spend $30-40, you might as well just buy a Teensy (v3.6) which comes with [a working MIDI Host USB Stack](https://github.com/PaulStoffregen/USBHost_t36).

My sad conclusion is that it may be a long time before TinyUSB gains MIDI Host support.

I have started looking at chips such as the CH559/MAX3421E/etc. but I've yet to find a working (and reasonably priced) solution.

**Reply by hathach** [reply]
host controller driver is not that hard to write, I just don't have the time to work on it. Plus 90% stack user use it as device so it makes more sense to spend more time on the device stack. Come back and check this project after several months to see if there is anything new.

**Reply by rppicomidi** [reply]
@hathach  I am also working on a host class driver for MIDI on the Raspberry Pi Pico. My code handles plug and unplug and successfully enumerates a a device. I am currently working on the bulk transfer implementation. Do you have any quick advice or function API you have in mind for the RP2040 USB Host bulk endpoints? Would you consider looking at a pull request when I am ready to start sharing code? I know that @csBlueChip already submitted a pull request for this, but I was already pretty far along when I noticed this and seem to have more working than what is in the pull request #1122 .

**Reply by hathach** [reply]
thanks @rppicomidi I plan to review your plug/unplug soon enough (currently busy with other paid works). 

For the host bulk endpoints, it is also carried out by `hcd_edpt_xfer()` similar to interrupt, the special case of rp2040 should be handled internally by its hcd/usb_rp2040. I have no guideline on the specific implementation, though you don't have to wait until the PR is fully functional to submit. Just submit it as draft one if you need to discuss more on the code and approach. #1122 is currently stalled, don't worry for the conflict, we could cherry-pick what is best from each PRs.

**Reply by csBlueChip** [reply]
@rppicomidi ...After being told the RP2040 is incapable of being a MIDI Host, and now Espressif have added native USBh support, I am out of this game ...I should probably withdraw the pull request - but there's loads of great reference URLs and parameter tables that you (or others) might like to utilise [all MIT - take, enjoy] ...probably the most valuable thing is the linux kernel patch with a list of VID/PIDs that lie in their descriptors.

I'll keep an eye on this thread, so feel free to fire questions ...please announce here if you get it working :)

**Reply by rppicomidi** [reply]
@csBlueChip  I have pulled down your fork of tinyusb. Thank you for pointing out where to look for stuff.

Perhaps what follows belongs in a new discussion thread or belongs in a forum for the Pico outside of this one. @hathach, let me know and I will act accordingly.

The current status is of my RP2040 MIDI host work is both devices I have tried enumerate correctly (Yamaha Reface CS, Korg nanoKONTROL Studio) I have the OUT endpoint working as well as I need for my application, but my application is not demanding (just send MIDI messages to the device on a button press). 

@hathach I am struggling with the IN endpoint. The Korg device I use for most of my testing should be sending a NAK for each transfer request when no controls are touched, but the RP2040 is happily reporting completed transfer for each transfer request. The buffers contain all 0s. I don't have a bus analyzer so I don't know whether the Korg test device or my code is at fault there. Also not clear is what to do with the  NAK_POLL register at offset 0x6c; I would think that a bulk IN endpoint that shares hardware with bulk OUT should not automatically retry on NAK, nor should it time out. It should set the NAK flag in the SIE status register, but I don't observe that. Do you know of a support path from the Raspberry Pi foundation where I can get direct answers to questions about the USB on the RP2040?

Finally, there are a number of random ways to make the system crash on unplug events or re-plug events that are separate from pull request #1193; the current host code can hit asserts if the device is detached from the host at the wrong time. I fixed one of those (crash if SIE status reports device speed 0, which means unplugged).

@hathach I will take the time to format a pull request after you have reviewed #1193 because the RP2040 MIDI host code needs those changes too (or something similar in case those changes are wrong).

**Reply by raveslave** [reply]
curious if this midi host code was completed

**Comment by rppicomidi** [comment]
@raveslave Yes it was. See pull request #1219 and discussion #1261. There is a newer pull request that uses the tinyusb master branch RP2040 Host Bulk Transfer method: #1627. I have a lot of examples on my GitHub page, including https://github.com/rppicomidi/midi2usbhost, https://github.com/rppicomidi/midi2usbhub and https://github.com/rppicomidi/midi2usbhub-web
